src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -217,4 +559,105 @@ mod tests {
+        {
+            let host_configuration = supported_cpu_config();
+
+            let template = X86_64CpuTemplate {
+                cpuid_modifiers: vec![],
+                msr_modifiers: vec![],
+            };
+
+            let guest_config_result = create_guest_cpu_config(&template, &host_configuration);
+            assert!(
+                guest_config_result.is_ok(),
+                "{}",
+                guest_config_result.unwrap_err()
+            );
+            assert_eq!(guest_config_result.unwrap(), host_configuration);
+        }
+
+        // TODO - Aarch64 test
+        // #[cfg(target_arch = "aarch64")]
+    }
+
+    #[test]
+    fn test_template() {
+        #[cfg(target_arch = "x86_64")]
+        {
+            let host_configuration = supported_cpu_config();
+
+            let template = X86_64CpuTemplate {
+                cpuid_modifiers: Vec::from([CpuidLeafModifier {
+                    leaf: 0x1,
+                    subleaf: 0x1,
+                    flags: KvmCpuidFlags(KVM_CPUID_FLAG_STATEFUL_FUNC),
+                    modifiers: Vec::from([
+                        CpuidRegisterModifier {
+                            register: CpuidRegister::Eax,
+                            mask: BitMask {
+                                filter: 0,
+                                value: 0,
+                            },
+                        },
+                        CpuidRegisterModifier {
+                            register: CpuidRegister::Eax,
+                            mask: BitMask {
+                                filter: 0,
+                                value: 0,
+                            },
+                        },
+                    ]),
+                }]),
+                msr_modifiers: Vec::from([RegisterModifier {
+                    addr: 0x8000,
+                    mask: BitMask {
+                        filter: 0,
+                        value: 0,
+                    },
+                }]),
+            };
+
+            let guest_config_result = create_guest_cpu_config(&template, &host_configuration);
+            assert!(
+                guest_config_result.is_ok(),
+                "{}",
+                guest_config_result.unwrap_err()
+            );
+            assert_ne!(guest_config_result.unwrap(), host_configuration);
+        }
+
+        // TODO - Aarch64 test
+        // #[cfg(target_arch = "aarch64")]
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn supported_cpu_config() -> X86_64CpuConfiguration {
+        let kvm_result = Kvm::new();
+        assert!(kvm_result.is_ok(), "Unable to access KVM");
+
+        // Create descriptor KVM resource's file descriptor
+        let vm_fd_result = kvm_result.as_ref().unwrap().create_vm();
+        assert!(vm_fd_result.is_ok(), "{}", vm_fd_result.unwrap_err());
+
+        let kvm_cpuid_result = kvm_result
+            .unwrap()
+            .get_supported_cpuid(KVM_MAX_CPUID_ENTRIES);
```

Comment:
I'm not really a fan of relying on KVM in unit tests. Can we create `X86_64CpuConfiguration` by hand, so it is determininstic?

Reply:
So the current `Cpuid` implementation is strongly tied to the vendor and building by hand is quite cumbersome and even trickier to debug if there is an issue. We do have hard-coded templates that I am fairly sure will do the job, as long as we don't try to use this for anything more than unit tests (the hard coded templates are Intel specific). I will swap this out to remove the KVM calls and see how it goes.